torch/csrc/api/include/torch/detail/TensorDataContainer.h

```diff
@@ -0,0 +1,272 @@
+    if (is_scalar()) {
+      at::AutoNonVariableTypeMode non_var_type_mode(true);
+      return at::scalar_tensor(scalar_, options.is_variable(false));
+    } else if (is_init_list()) {
+      // NOTE: Here we explicitly choose to initialize the tensor on CPU first,
+      // fill each element of the tensor, and then move the tensor to the desired
+      // device. For CUDA device, this approach only involves 1 CUDA kernel launch,
+      // and is much faster than initializing the tensor on CUDA first and then
+      // filling each element of it (which involves `N` CUDA kernel launches where
+      // `N` is the number of the elements in the tensor).
+      at::Tensor tensor = ([&]() {
+        at::AutoNonVariableTypeMode non_var_type_mode(true);
+        return at::empty(sizes_, options.device(at::kCPU).is_variable(false));
+      })();
+      fill_tensor(tensor);
+      return tensor.to(options.device());
+    } else if (is_tensor()) {
+      return tensor_.to(options);
+    } else {
+      TORCH_INTERNAL_ASSERT(false, "Invalid TensorDataContainer type");
+    }
+  }
+
+  void pretty_print_recursive(std::ostream& stream) const {
+    if (is_scalar()) {
+      // NOTE: There is no `operator<<` overload for `at::kBFloat16` type,
+      // and we need to convert it to `float` type using `operator float()` function
+      // defined in c10/util/BFloat16.h.
+      if (scalar_type_ == at::ScalarType::BFloat16) {
+        stream << static_cast<float>(scalar_.to<c10::BFloat16>());
+      } else {
+        AT_DISPATCH_ALL_TYPES_AND2(at::kBool, at::kHalf, scalar_type_, "TensorDataContainer_pretty_print_scalar", [&] {
+          stream << scalar_.to<scalar_t>();
+        });
+      }
+    } else if (is_init_list()) {
+      stream << "{";
+      for (const TensorDataContainer* it = init_list_.begin(); it != init_list_.end(); it++) {
+        stream << *it;
+        if (std::next(it) != init_list_.end()) stream << ", ";
+      }
+      stream << "}";
+    } else if (is_tensor()) {
+      stream << "{";
+      for (int64_t i = 0; i < tensor_.sizes()[0]; i++) {
+        // NOTE: There is no `operator<<` overload for `at::kBFloat16` type,
+        // and we need to convert it to `float` type using `operator float()` function
+        // defined in c10/util/BFloat16.h.
+        if (scalar_type_ == at::ScalarType::BFloat16) {
+          stream << static_cast<float>(tensor_[i].item<c10::BFloat16>());
+        } else {
+          AT_DISPATCH_ALL_TYPES_AND2(at::kBool, at::kHalf, scalar_type_, "TensorDataContainer_pretty_print_tensor_item", [&] {
+            stream << tensor_[i].item<scalar_t>();
+          });
+        }
+      }
+      stream << "}";
+    } else {
+      TORCH_INTERNAL_ASSERT(false, "Invalid TensorDataContainer type");
+    }
+  }
+ private:
+  void fill_tensor(at::Tensor tensor) const {
+    if (is_scalar()) {
+      at::NoGradGuard guard;
+      tensor.fill_(scalar_);
+    } else if (is_init_list()) {
```

Comment:
We do no error checking in this function -- is the assumption that `tensor` is always the correct size? If so, could we add some asserts?

Reply:
Thanks for the suggestion! I added asserts to this function